server/tailsql/options.go

```diff
@@ -81,26 +85,43 @@ type Options struct {
 	Logf logger.Logf `json:"-"`
 }
 
-// Construct database handles to serve queries from. This returns nil without
-// error if no sources are defined.
-func (o Options) sources() ([]*dbHandle, error) {
+// openSources opens database handles to each of the sources defined by o.
+// Sources that require secrets will get them from store.
+// Precondition: All the sources of o have already been validated.
+func (o Options) openSources(store *setec.Store) ([]*dbHandle, error) {
 	if len(o.Sources) == 0 {
 		return nil, nil
 	}
+
 	srcs := make([]*dbHandle, len(o.Sources))
 	for i, spec := range o.Sources {
-		if err := spec.checkValid(); err != nil {
-			return nil, err
-		} else if spec.Label == "" {
+		if spec.Label == "" {
 			spec.Label = "(unidentified database)"
 		}
 
-		db, err := sql.Open(spec.Driver, spec.URL)
+		// Resolve the connection string.
+		var connString string
+		switch {
+		case spec.URL != "":
+			connString = spec.URL
+		case spec.KeyFile != "":
+			data, err := os.ReadFile(os.ExpandEnv(spec.KeyFile))
+			if err != nil {
+				return nil, fmt.Errorf("read key file for %q: %w", spec.Source, err)
+			}
+			connString = string(data)
+		case spec.Secret != "":
+			connString = string(store.Secret(spec.Secret).Get())
```

Comment:
I assume tailsql never reopens the DB after startup? So it's moot that it's not trying to refresh when the secret changes?

Reply:
That's right, we just create and hold open the connection for the life of the process. (It makes connections underneath, but they all share the same connection string).